ext/msgpack/packer.c

```diff
@@ -131,45 +134,39 @@ void msgpack_packer_write_other_value(msgpack_packer_t* pk, VALUE v)
         VALUE payload = rb_funcall(proc, s_call, 1, v);
         StringValue(payload);
         msgpack_packer_write_ext(pk, ext_type, payload);
+    } else if(pk->strict_types) {
+        VALUE exc = rb_obj_alloc(ePackError);
+        rb_ivar_set(exc, rb_intern("@error_value"), v);
+        rb_exc_raise(exc);
     } else {
         rb_funcall(v, pk->to_msgpack_method, 1, pk->to_msgpack_arg);
     }
 }
 
 void msgpack_packer_write_value(msgpack_packer_t* pk, VALUE v)
 {
-    switch(rb_type(v)) {
-    case T_NIL:
+    VALUE klass = rb_class_of(v);
```

Comment:
I don't understand how this works. Previously a `HashWithIndifferentAccess` would end up in `msgpack_packer_write_hash_value`, now it end in `msgpack_packer_write_other_value` ?

So they are being serailized as extension types? I'm confused.

Reply:
Yes exactly, with `strict: true` they're being treated as extension types rather than treated as the core type. The corresponding change in `msgpack_packer_write_other_value` means that if it finds no extension type, rather than try to call `to_msgpack` it just raises a `MessagePack::PackError`. This gets around the problem that all subclasses of core types have `to_msgpack` defined on them because they inherit from the parent core type. It also has the nice side-effect that it raises an exception that we can actually catch nicely, rather than `NoMethodError`.

Reply:
My point is that with `strict:false` it would change the serialization format, no?

Reply:
Yes, I saw that test, that's why I'm confused. If subclasses fallback to extension types, I would have expected the type code to change.

Reply:
Added an explicit test that encoded values don't change with `strict: false` in 9b8e7c74070eaf62b7f1102d8d18588b2ba03a6b